### PR TITLE
Versioning

### PR DESCRIPTION
**Pull request by dannylamb**
All work done by @elizoller.  Can't take any credit on this one.  I just migrated this PR over from https://github.com/Islandora-CLAW/islandora/pull/154

**GitHub Issue**: Related to: (https://github.com/Islandora-CLAW/CLAW/issues/740)

# What does this Pull Request do?
* Allows versioning of an object to be initiated from Drupal node updates

# What's new?
* Adds a context action create_node_version_in_fedora which can be added to a context reaction.
* Adds the event type to the event object being passed to Alpaca
* Adds a "Version" event type
* This change could be added to the index "reaction" in the repository item context by default and/or added to documentation that this is how one "enables" versioning on a node

# How should this be tested?
* pull in the PR to your islandora modules - uninstall/re-enable islandora_core_feature
* add the "Create Node Version in Fedora" to the "Index" reaction in the repository_content context
* make sure that the "Version" event gets fired on node_update (could probably just run a debug)
* make sure that the event type is included in the emitted event object (could debug from EventGenerator or update an alpaca feature to try this)

# Additional Notes:
One important note is that using the IndexReaction for triggering the create_node_version_in_fedora means that the event is triggered as part of islandora_node_insert and islandora_node_update (i think...). Having it be part of node_update means that the version is created before the node is updated. 
Another way to do this is to use a new reaction (maybe called VersionReaction) that could then be included in any hooks in `islandora.module` although probably islandora_node_update makes the most sense.

# Interested parties
@Islandora-CLAW/committers

### CODE REVIEW

**Comment by elizoller** [comment]
I can't edit this PR, but an update on the instructions:
You just need to pull down this PR as your islandora module, there is no configuration necessary as the versioning is fired automatically on node by the EventGenerator on bundles which are versionable and return true with shouldCreateNewRevision

**Comment by seth-shaw-unlv** [comment]
@elizoller build failed due to a coding standard issue. (An excessively long comment, it appears.)

File: src/EventGenerator/EventGenerator.php [hunk 1]
```diff
@@ -145,4 +154,42 @@ public function generateEvent(EntityInterface $entity, UserInterface $user, arra
...
+   *   Media object.
+   * @param \Drupal\Core\Entity\EntityStorageInterface $media_storage
+   *   Media Storage.
+   */
+  protected function getRevisionIds(Media $media, EntityStorageInterface $media_storage) {
+    $result = $media_storage->getQuery()
+      ->allRevisions()
+      ->condition($media->getEntityType()->getKey('id'), $media->id())
+      ->sort($media->getEntityType()->getKey('revision'), 'DESC')
+      ->pager(50)
```
Comment (seth-shaw-unlv) [inline src/EventGenerator/EventGenerator.php]: While using pager is fine with how we use the function now (we only check if there is more than one); if someone re-uses the function call and does need all of the revision ids they will only get the most recent 50 resulting, potentially, in a hidden logic error for media with 50+ revisions.

File: src/EventGenerator/EventGenerator.php [hunk 2]
```diff
@@ -98,6 +100,13 @@ public function generateEvent(EntityInterface $entity, UserInterface $user, arra
       $event["type"] = ucfirst($data["event"]);
       $event["summary"] = ucfirst($data["event"]) . " a " . ucfirst($entity_type);
     }
+    $isNewRev = FALSE;
+    if ($entity->getEntityType()->isRevisionable()) {
+      \Drupal::logger('islandora')->notice('its revisionable');
```
[the remark this answers is not in the document]
Comment (elizoller) [reply src/EventGenerator/EventGenerator.php]: should be fixed with 0f91ce5

File: src/EventGenerator/EventGenerator.php [hunk 3]
```diff
@@ -98,6 +100,13 @@ public function generateEvent(EntityInterface $entity, UserInterface $user, arra
...
+      $isNewRev = $this->isNewRevision($entity);
+      \Drupal::logger('is new revision')->notice($isNewRev);
```
[the remark this answers is not in the document]
Comment (elizoller) [reply src/EventGenerator/EventGenerator.php]: should be fixed with 0f91ce5

File: src/EventGenerator/EventGenerator.php (hunk 1 above)
Comment (elizoller) [reply src/EventGenerator/EventGenerator.php]: should be fixed with 0f91ce5

**Comment by seth-shaw-unlv** [comment]
This looks good; but I'll hold off on merging until the related PRs are green.

**Pull request merged**